### DISCUSSION TITLE
[Paid support] Set up per tenant authentication for Orchard Core Framework app

### DISCUSSION THREAD

**Discussion by xperiandri** [OrchardCMS/OrchardCore#10423] [General]
[Original discussion](https://github.com/OrchardCMS/OrchardCore/discussions/10082)

We use Orchard Core Framework, without any CMS modules.
We need to set up authentication via Bitrix 24 OAuth2 ASP.NET Core authentication provider.
The provider must use different client id, client secret, and endpoint URLs with tenant hostname in it.

When we tried to set it up the same as in regular ASP.NET Core app, the Bitrix 24 OAuth2 provider is created for the first requested tenant and is reused for any other tenant initialized afterwards of course causing authentication error.
Reproduction sample is here https://github.com/maglini/multi-tenancy-app

When we tried to set it up similar to Users module (registering all the stuff within a module ourselves) it does not work

**Web app**

``` C#
services.AddOrchardCore()
    .AddMvc()
    .WithTenants()
    .AddAzureTableShellsConfiguration()
    .ConfigureServices(s =>
    {
        s.AddRazorPages();
        s.AddSingleton<IShellHost, ShellHost>();
        s.AddSingleton<IShellDescriptorManagerEventHandler>(sp => sp.GetRequiredService<IShellHost>());
        s.AddSingleton<ITenantService, TenantService>();
        s.AddAuthentication(options => options.DefaultSignOutScheme = IdentityConstants.ApplicationScheme);
    })
    .Configure((app, routes, srv) =>
    {
        app.UseAuthorization();
        app.UseWelcomePage("/");
    });
```

**Module**

``` C#
services.ConfigureApplicationCookie(options =>
{
	var tenantName = shellSettings.Name;
	options.Cookie.Name = "orchauth_" + HttpUtility.UrlEncode(tenantName);
	options.LoginPath = $"/signin";
	options.LogoutPath = $"/signout";
});

services.TryAddEnumerable(
	ServiceDescriptor.Singleton<
		IPostConfigureOptions<
			Bitrix24AuthenticationOptions>,
			OAuthPostConfigureOptions<Bitrix24AuthenticationOptions, TenantBitrix24AuthenticationHandler>>());

services.TryAddSingleton<IPostConfigureOptions<Bitrix24AuthenticationOptions>, Bitrix24PostConfigureOptions>();
services.AddTransient<TenantBitrix24AuthenticationHandler>();
services.Configure<AuthenticationOptions>(options =>
{
	options.AddScheme<TenantBitrix24AuthenticationHandler>(
		Bitrix24AuthenticationDefaults.AuthenticationScheme,
		displayName: Bitrix24AuthenticationDefaults.DisplayName);
});
services.Configure<Bitrix24AuthenticationOptions>(Bitrix24AuthenticationDefaults.AuthenticationScheme, options => { });
```

Here we get 404
`No webpage was found for the web address: https://localhost:44318/oauth/authorize/?client_id=&scope=&response_type=code&redirect_uri=https%3A%2F%2Flocalhost%3A44318%2Fartsoft.bitrix24.ru%2Fsignin-bitrix24&state=&response_mode=form_post`

We tried to prepend the URL with tenant name with no luck
https://localhost:44318/artsoft.bitrix24.ru/oauth/authorize/?client_id=&scope=&response_type=code&redirect_uri=https%3A%2F%2Flocalhost%3A44318%2Fartsoft.bitrix24.ru%2Fsignin-bitrix24&state=&response_mode=form_post

Help needed

**Comment by Skrypt** [comment]
Maybe you can ask @kevinchalet he's the expert with OpenAuth. Else, if the issue is how you register these services in tenants then the expert is @jtkech 

I can't say I have time to help you neither that I have the experience with Bitrix so it's an unknown territory for everyone.

**Comment by ns8482e** [comment]
@xperiandri  If you are facing issue resolving to tenant prefix after authentication in dev then use HostName instead of `RequestUrlPrefix`

e.g. If you have following setup for prod vs development 

|PROD | DEV|
|------|---|
|www.site1.com |  localhost:5000/site1 |
|www.site2.com |  localhost:5000/site2 |

Then to update your local environment like prod - Update your hosts file as following
127.0.0.1   www.site1.com
127.0.0.1   www.site2.com

and in OrchardCore keep `RequestUrlPrefix` empty and provide host name for each tenant.

**Reply by Skrypt** [reply]
Good one. Get rid of the tenant prefixes with your DNS host file.

**Reply by Skrypt** [reply]
Also does the TenantBitrix24AuthenticationHandler requires different settings per tenant? Because here you are registering the same handler for each tenants. I'm not sure how this handler works.

**Reply by Skrypt** [reply]
You must know that one already I'm guessing : https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/blob/0987305a6b3440d22a6d62fc424a4f9f1c6e778b/docs/bitrix24.md

**Reply by ns8482e** [reply]
Validate your returnurl/replyurl from Bitrix24 it should return to correct tenant after authentication

**Comment by xperiandri** [comment]
So the issue was on our side. We forgot to set domain name from `ShellSettings.Name` property to Bitrix 24 identity provider.
Now it works with the original setup (i the sample). Not the code published here.
